feat(baselines) Add memory-based TraceFL provenance integration for federated learning

ibrahim-Cypher10 opened the pull request:
## Issue

### Description

This PR contributes a Flower-compatible implementation of **TraceFL**, a provenance analysis tool for federated learning. TraceFL was originally proposed in the paper [TraceFL: A Provenance Framework for Federated Learning](https://arxiv.org/abs/2312.13632).

## Proposal

### Explanation

The original implementation of TraceFL relied on cache-based storage and was not integrated with the Flower ecosystem. This PR converts TraceFL into a fully in-memory, non-cached version that integrates directly with Flower’s federated learning workflow. 

It enables provenance analysis during and after training rounds by tracking neuron activations and gradients in memory. The conversion enhances usability and compatibility while preserving the original tool’s core functionality.

For detailed documentation and original implementation, refer to the official TraceFL repository: [https://github.com/SEED-VT/TraceFL](https://github.com/SEED-VT/TraceFL).

### Checklist

- [x] Integrated TraceFL into Flower as a memory-based tool
- [x] Removed all disk- or cache-based dependencies
- [x] Ensured compatibility with Flower strategies and client config
- [x] Passed local tests and verified client-to-class mappings
- [x] Prepped for baseline inclusion

### Any other comments?

Looking forward to feedback from maintainers. Happy to make adjustments based on review!

ibrahim-Cypher10 committed:
feat(Baseline) Tracefl

WilliamLindskog commented:
linking this to #5245

ibrahim-Cypher10 committed:
Final TraceFL Baseline Update – Code Clean, Tests Pass, and README Added

Thought for a couple of seconds

All checks succeed with `./dev/format-baseline.sh` and `./dev/test-baseline.sh`, and a clear, easy‑to‑follow README is now included.

jafermarq commented on baselines/tracefl/pyproject.toml:
```diff
@@ -0,0 +1,185 @@
+  "omegaconf==2.3.0",
+  
+  # Utilities
+  "hydra-core==1.3.2",
```
```suggestion
```

jafermarq replied:
I believe Hydra isn't being used, therefore we can remove it

ibrahim-Cypher10 replied:
Removed

jafermarq commented on baselines/tracefl/tracefl/client_app.py:
```diff
@@ -0,0 +1,169 @@
+
+    client_train_data = ds_dict["client2data"].get(str(partition_id))
+
+    sim = FLSimulation(copy.deepcopy(cfg), 0.0, 0.0, 0.0, 0.0)
+    sim.set_clients_data(ds_dict["client2data"])
+    sim.set_strategy()
```
what is this for ? `sim` doesn't seem to be used after these lines.

jafermarq commented on baselines/tracefl/README.md:
```diff
@@ -0,0 +1,130 @@
+---
+
+title: TraceFL: Interpretability-Driven Debugging in Federated Learning via Neuron Provenance
+url: [https://arxiv.org/abs/2312.13632](https://arxiv.org/abs/2312.13632)
+labels: \[interpretability, debugging, federated-learning, neuron-provenance, explainability, computer-vision, nlp, dirichlet-partitioning, client-attribution, auditing]
+dataset: \[MNIST, CIFAR-10, PathMNIST, YahooAnswers]
+------------------------------------------------------------------
```
```suggestion
---
title: TraceFL: Interpretability-Driven Debugging in Federated Learning via Neuron Provenance
url: https://arxiv.org/abs/2312.13632
labels: [interpretability, debugging, federated-learning, neuron-provenance, explainability, computer-vision, nlp, dirichlet-partitioning, client-attribution, auditing]
dataset: [MNIST, CIFAR-10, PathMNIST, YahooAnswers]
---
```

jafermarq replied:
formatting wasn't correct. Could you also choose 3 or 4 among those `labels`?

ibrahim-Cypher10 replied:
Fixed

jafermarq commented on baselines/tracefl/README.md:
```diff
@@ -0,0 +1,130 @@
+> \[!NOTE]
+> If you use this baseline in your work, please remember to cite the original authors of the paper as well as the Flower paper.
+
+**Paper:** [TraceFL: Interpretability-Driven Debugging in Federated Learning via Neuron Provenance](https://arxiv.org/abs/2312.13632)
```
```suggestion
**Paper:** [https://arxiv.org/abs/2312.13632](https://arxiv.org/abs/2312.13632)
```

ibrahim-Cypher10 replied:
Addressed in the latest commit

jafermarq commented on baselines/tracefl/README.md:
````diff
@@ -0,0 +1,130 @@
+```bash
+flwr run . --run-config exp_2
+```
+
````
Based on the _"what's implemented"_ paragraph at the top of the readme, here i'd expect to see Figs 2,3,5 and table 3 with the results the code in the PR generates.

ibrahim-Cypher10 replied:
Real implementation results have been generated and included:
- Added "Expected Results" section showing real performance metrics
- Added "Reproducing Results" section with script instructions
- All experiment sections now show actual implementation performance results

jafermarq commented on baselines/tracefl/README.md:
```diff
@@ -0,0 +1,130 @@
+
+## Running the Experiments
+
+### Experiment 1 — Correct Prediction Tracing 
```
Does this mean the command below will generate results for Figure 2,5 and Table 3?

ibrahim-Cypher10 replied:
Addressed in the new pushed Readme file

jafermarq commented on baselines/tracefl/tracefl/utils.py:
```diff
@@ -0,0 +1,126 @@
+    return max(items, key=scoring_fn)
+
+
+def get_backend_config(cfg):
```
I'm intrigued. What is this function used for?

jafermarq commented on baselines/tracefl/tracefl/fls.py:
```diff
@@ -0,0 +1,215 @@
+from tracefl.utils import get_backend_config
+
+
+class FLSimulation:
```
What is the main purpose of this class ?

ibrahim-Cypher10 replied:
The FLSimulation class is essential and serves as the central orchestrator for TraceFL federated learning experiments. Here's why:
1. **Strategy Configuration**: Creates and configures the appropriate FL strategy based on configuration (with or without differential privacy)
2. **Model Initialization**: Handles model setup and parameter initialization with proper validation
3. **Data Management**: Sets up server and client data for provenance analysis
4. **Provenance Integration**: Integrates neuron provenance tracking capabilities into the FL workflow
5. **Evaluation Orchestration**: Manages global model evaluation and metric tracking
6. **DP Integration**: Conditionally wraps strategies with differential privacy when enabled

jafermarq reviewed (commented):
Hello @ibrahim-Cypher10, thanks for opening this PR for `TraceFL`! I did a first pass through the code but I'm a bit confused about the `FLSimulation` class and all the functionality built around it. What is it for and is it used? (1) in the `ClientApp` it's defined but not used from what I can tell; (2) in the `ServerApp` it's defined but then only the strategy is used (which is passed to the return statement of `server_fn`). So i'm wondering if we can remove `FLSimulation`.

I also left other comments below. Likely the most important is that it's not clear for a reader going through the readme what would be the results (plots/figures) to be able to generate when running each of the commands you indicate later. I suggested looking at how for example [FedProx](https://github.com/adap/flower/tree/main/baselines/fedprox) or [FedRep](https://github.com/adap/flower/tree/main/baselines/fedrep) or [StatAvg](https://github.com/adap/flower/tree/main/baselines/statavg) -- or others.

ibrahim-Cypher10 commented:
Thank you @jafermarq for pointing these out. This is really helpful, and I understand your concerns. I’ll work on the fixes and update soon.

WilliamLindskog commented:
Hi @ibrahim-Cypher10, 

just checking in here to see if you're working on the changes. We'd be happy to review new changes.

ibrahim-Cypher10 committed:
TraceFL Cleanup: Removed unused FLSimulation, updated README, cleaned models.py, removed Hydra

Final fixes for TraceFL baseline:
1. Removed unused FLSimulation references
2. Models.py is cleaned up and broken into 3 smaller files
3. Removed Hydra from dependencies
4. Fixed README formatting and added expected results section

ibrahim-Cypher10 committed:
Update README.md

ibrahim-Cypher10 committed:
readme fix

ibrahim-Cypher10 commented on baselines/tracefl/README.md:
````diff
@@ -0,0 +1,130 @@
+}
+```
+
+---
+
+> Maintained as part of the Flower Baselines Collection
````
Fixed

ibrahim-Cypher10 commented on baselines/tracefl/tracefl/models.py:
Agreed, but since it's essential, I've split it into three files to improve readability.

ibrahim-Cypher10 committed:
Update README.md

ibrahim-Cypher10 committed:
removed get_backend_config

ibrahim-Cypher10 commented:
Hi @WilliamLindskog,

I've replied to all the conversation threads. Let me know if there's anything else needed.

Best regards,
Ibrahim

ibrahim-Cypher10 commented:
Hi @WilliamLindskog @jafermarq , I hope you’re doing well. It has been some time since the code was committed, I just wanted to kindly check if there’s anything further needed from my side or if the PR is all set (sharing this as a gentle reminder in case it was overlooked).

ibrahim-Cypher10 committed:
Merge branch 'main' into main

ibrahim-Cypher10 committed:
Merge branch 'main' into main

ibrahim-Cypher10 committed:
Merge branch 'adap:main' into main

ibrahim-Cypher10 committed:
tracefl

ibrahim-Cypher10 committed:
traefl

ibrahim-Cypher10 committed:
lint

ibrahim-Cypher10 committed:
cons

ibrahim-Cypher10 committed:
cons

ibrahim-Cypher10 committed:
class fix

ibrahim-Cypher10 committed:
scripts fix

ibrahim-Cypher10 committed:
neuron prov

ibrahim-Cypher10 committed:
deterministic mode

ibrahim-Cypher10 committed:
deterministic mode

ibrahim-Cypher10 committed:
deterministic mode

ibrahim-Cypher10 committed:
deterministic mode

ibrahim-Cypher10 committed:
dirich-0.3

ibrahim-Cypher10 committed:
dirchc ons

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
fix

ibrahim-Cypher10 committed:
fix

ibrahim-Cypher10 committed:
fix

ibrahim-Cypher10 committed:
log repo

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
dp

ibrahim-Cypher10 committed:
expt

ibrahim-Cypher10 committed:
with readme

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
fix graph

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
utils

ibrahim-Cypher10 committed:
comment fix

ibrahim-Cypher10 committed:
Readme for replication results

ibrahim-Cypher10 committed:
dep

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
graph for diff pr

ibrahim-Cypher10 committed:
unneeded fallback

ibrahim-Cypher10 committed:
config of faulty clients

ibrahim-Cypher10 committed:
config of faulty clients

ibrahim-Cypher10 committed:
conf

ibrahim-Cypher10 committed:
c expt

ibrahim-Cypher10 committed:
c expt

ibrahim-Cypher10 committed:
faulty

ibrahim-Cypher10 committed:
faulty

ibrahim-Cypher10 committed:
all expts

ibrahim-Cypher10 committed:
fix

ibrahim-Cypher10 committed:
experiments rerun

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
Final push

ibrahim-Cypher10 committed:
readme final

ibrahim-Cypher10 committed:
readme typo

ibrahim-Cypher10 committed:
readme fix regarding version

ibrahim-Cypher10 committed:
readme

ibrahim-Cypher10 committed:
Test passed

ibrahim-Cypher10 committed:
Merge branch 'adap:main' into main

ibrahim-Cypher10 committed:
Merge branch 'main' into main

jafermarq commented:
Hi @ibrahim-Cypher10 , could we discard the `results/` directory? there are a few large files in there.

ibrahim-Cypher10 committed:
Stop tracking results folder
